[mlir][spirv] Add folding for [S|U]Mod, [S|U]Div, SRem

inbelic opened the pull request:
Add missing constant propogation folder for [S|U]Mod, [S|U]Div, SRem

Implement additional folding when rhs is 1 for all ops.

This helps for readability of lowered code into SPIR-V.

Part of work for #70704

inbelic committed:
[mlir][spirv] Add folding for [S|U]Mod, [S|U]Div, SRem

Add missing constant propogation folder for [S|U]Mod, [S|U]Div, SRem

Implement additional folding when rhs is 1 for all ops.

This helps for readability of lowered code into SPIR-V.

Part of work for #70704

kuhar commented on mlir/lib/Dialect/SPIRV/IR/SPIRVCanonicalization.cpp:
```diff
@@ -69,6 +69,14 @@ static Attribute extractCompositeElement(Attribute composite,
 }
 
+static bool isDivZeroOrOverflow(const APInt &a, const APInt &b) {
+  bool div0 = b.isZero();
+
+  bool overflow = a.isMinSignedValue() && b.isAllOnes();
+
```
```suggestion
  bool div0 = b.isZero();
  bool overflow = a.isMinSignedValue() && b.isAllOnes();
```

kuhar commented on mlir/lib/Dialect/SPIRV/IR/SPIRVCanonicalization.cpp:
```diff
@@ -290,6 +298,158 @@ OpFoldResult spirv::ISubOp::fold(FoldAdaptor adaptor) {
+  // result is the remainder r of Operand 1 divided by Operand 2 where if
+  // r ≠ 0, the sign of r is the same as the sign of Operand 2.
+  //
+  // So don't fold during undefined behaviour
```
```suggestion
  // So don't fold during undefined behavior.
```

kuhar commented on mlir/test/Dialect/SPIRV/Transforms/canonicalize.mlir:
```diff
@@ -462,10 +462,272 @@ func.func @const_fold_vector_isub() -> vector<3xi32> {
+
+// CHECK-LABEL: @sdiv_div_0_or_overflow
+func.func @sdiv_div_0_or_overflow() -> (i32, i32) {
+  // CHECK: spirv.SDiv
+  // CHECK: spirv.SDiv
```
Could we match the operands to confirm that they remain the same?

kuhar commented on mlir/test/Dialect/SPIRV/Transforms/canonicalize.mlir:
```diff
@@ -462,10 +462,272 @@ func.func @const_fold_vector_isub() -> vector<3xi32> {
+  %c3 = spirv.Constant 3 : i32
+  %cn3 = spirv.Constant -3 : i32
+
+  // CHECK-DAG: spirv.Constant -18
+  // CHECK-DAG: spirv.Constant -2
+  // CHECK-DAG: spirv.Constant -7
+  // CHECK-DAG: spirv.Constant 8
```
Here and everywhere else: we should make sure that these constants are used correctly. In this instance we should match the returned values to make sure they are in the correct order.

kuhar commented on mlir/test/Dialect/SPIRV/Transforms/canonicalize.mlir:
```diff
@@ -462,10 +462,272 @@ func.func @const_fold_vector_isub() -> vector<3xi32> {
+  return %2 : i32
+}
+
+// CHECK-LABEL: @sdiv_div_0_or_overflow
+func.func @sdiv_div_0_or_overflow() -> (i32, i32) {
```
nice, thanks for testing this corner case!

kuhar reviewed (commented):
Awesome, thanks for adding these @inbelic! The implementation looks good to me, but I think we should make the tests match the expected results more precisely.

inbelic committed:
review comments

- fix spacing issue
- correct spelling
- make testcases more strict when matching the return values to ensure
  proper order

kuhar reviewed (approved):
LGTM, thanks for the contribution!
